### ISSUE TITLE
Long running streams fail

### ISSUE THREAD

**Issue by daveoxley** [grpc/grpc-go#7641]
### What version of gRPC are you using?

1.66.2
But seen on all versions of 1.66.x so far. Issue did not occur on 1.65

### What version of Go are you using (`go version`)?

N/A

### What operating system (Linux, Windows, …) and version?

N/A

### What did you do?

Long running stream fails with an error. Test case in https://github.com/grpc/grpc-go/pull/7642.

### What did you expect to see?

Stream completing successfully:
```
=== RUN   TestLongRunningStream
--- PASS: TestLongRunningStream (127.39s)
PASS
```

### What did you see instead?

Errors vary between the 2 below:
```
=== RUN   TestLongRunningStream
    stream_test.go:70: Failed to receive response: rpc error: code = Internal desc = received 4294967270-bytes data exceeding the limit 65535 bytes
--- FAIL: TestLongRunningStream (17.32s)
```
and
```
=== RUN   TestLongRunningStream
    stream_test.go:70: Failed to receive response: rpc error: code = Internal desc = unexpected EOF
--- FAIL: TestLongRunningStream (8.14s)
```

**Comment by arjan-bal** [comment]
After using [git bisect](https://git-scm.com/docs/git-bisect) on the range from tag `1.65.0-dev` to HEAD of master, the commit causing the regression seems to be https://github.com/grpc/grpc-go/commit/9ab8b62505c85a5e03be1ae97949e960e24b6d4f.

Command:
```sh
# Run 10 times in parallel for higher confidence
seq 1 10 | xargs -n1 -P10 -I{}  go test ./test/stream/ -run TestLongRunningStream -count 1 -v || echo "A test failed"
```

I then verified that the test flakes on the the commit and passed 30/30 times on the commit immediately before it.

**Comment by zhouyan** [comment]
I have been seeing the same problems as well since upgrade to 1.66 and beyond.

the error I got was control flow error, saying received data exceeding limit. The received data is around 4GB according to the logs.

**Comment by arjan-bal** [comment]
It seems like the `amount of bytes being read > the amount of bytes received`. This is causing an unsigned 32 bit integer to underflow resulting in 4294967270  which is -26.

I think I found the bug. The headers are being read incrementally here:
https://github.com/grpc/grpc-go/blob/8ea3460acace12cd1f2f0502784d42c21eeadf42/internal/transport/transport.go#L559-L579

We keep filling the underlying `header` slice until its full, but we do so by moving the start of the slice forward. In the following function that is called by the function above to perform reads and update the flow control window, we do `t.windowHandler(len(header))`.
https://github.com/grpc/grpc-go/blob/8ea3460acace12cd1f2f0502784d42c21eeadf42/internal/transport/transport.go#L622-L630

The code has read `n` bytes, but it says it read `len(header)` bytes. `len(header)` is the total number of bytes that remain to be read, which is not always equal the the bytes there were read. I was able to get the test to pass after changing the line to `t.windowHandler(n)`.
